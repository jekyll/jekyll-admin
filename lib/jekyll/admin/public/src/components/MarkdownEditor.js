import React, { Component } from 'react';
import SimpleMDE from 'simplemde-fork';

const classNames = [
  'editor-toolbar',
  'CodeMirror',
  'editor-preview-side',
  'editor-statusbar'
];

class MarkdownEditor extends Component {

  componentDidMount() {
    this.create();
  }

  componentDidUpdate() {
    this.destroy(); this.create();
  }

  componentWillUnmount() {
    this.destroy();
  }

  create() {
    let opts = Object.create(this.props);
    opts['element'] = this.refs.text;
    opts['hideIcons'] = ["guide"];
    opts['autoDownloadFontAwesome'] = false;
    opts['spellChecker'] = false;
    this.editor = new SimpleMDE(opts);
  }

  getValue() {
    return this.editor.value();
  }

  destroy() {
    for (let i in classNames) {
      let elementToRemove = this.refs.container.querySelector('.' + classNames[i]);
      elementToRemove && elementToRemove.remove();
    }
  }

  render() {
    return React.createElement('div', { ref: 'container' },
      React.createElement('textarea', { ref: 'text' })
    );
  }
}

export default MarkdownEditor;
