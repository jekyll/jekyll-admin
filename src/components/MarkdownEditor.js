import React, { Component } from 'react';
import PropTypes from 'prop-types';
import SimpleMDE from 'simplemde';
import hljs from '../utils/highlighter';

const classNames = [
  'editor-toolbar',
  'CodeMirror',
  'editor-preview-side',
  'editor-statusbar',
];

class MarkdownEditor extends Component {
  componentDidMount() {
    this.create();
    window.hljs = hljs; // TODO: fix this after the next release of SimpleMDE
  }

  shouldComponentUpdate(nextProps) {
    return nextProps.initialValue !== this.props.initialValue;
  }

  componentDidUpdate() {
    this.destroy();
    this.create();
  }

  componentWillUnmount() {
    this.destroy();
  }

  create() {
    const { onChange, onSave } = this.props;
    let opts = Object.create(this.props);
    opts['element'] = this.refs.text;
    opts['autoDownloadFontAwesome'] = false;
    opts['spellChecker'] = false;
    opts['renderingConfig'] = {
      codeSyntaxHighlighting: true,
    };
    let toolbarIcons = [
      'bold',
      'italic',
      'heading',
      '|',
      'code',
      'quote',
      'unordered-list',
      'ordered-list',
      '|',
      'link',
      'image',
      'table',
      '|',
      'preview',
      'side-by-side',
      'fullscreen',
      '|',
    ];
    if (onSave) {
      toolbarIcons.push({
        name: 'save',
        action: onSave,
        className: 'fa fa-floppy-o',
        title: 'Save',
      });
    }
    opts['toolbar'] = toolbarIcons;
    const editor = new SimpleMDE(opts);
    if (editor.codemirror) {
      editor.codemirror.on('change', () => {
        onChange(editor.value());
      });
    }
    this.editor = editor;
  }

  destroy() {
    for (let i in classNames) {
      let elementToRemove = this.refs.container.querySelector(
        '.' + classNames[i]
      );
      elementToRemove && elementToRemove.remove();
    }
  }

  render() {
    return React.createElement(
      'div',
      { ref: 'container' },
      React.createElement('textarea', { ref: 'text' })
    );
  }
}

MarkdownEditor.propTypes = {
  initialValue: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  onSave: PropTypes.func.isRequired,
};

export default MarkdownEditor;
