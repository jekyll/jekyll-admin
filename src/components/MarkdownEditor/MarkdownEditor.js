import React, { Component } from 'react';
import PropTypes from 'prop-types';
import SimpleMDE from 'simplemde';
import './font-awesome/css/font-awesome.min.css';
import 'simplemde/dist/simplemde.min.css';

const classNames = [
  'editor-toolbar',
  'CodeMirror',
  'editor-preview-side',
  'editor-statusbar',
];

class MarkdownEditor extends Component {
  componentDidMount() {
    this.create();
  }

  shouldComponentUpdate(nextProps) {
    return nextProps.initialValue !== this.props.initialValue;
  }

  componentDidUpdate() {
    this.destroy();
    this.create();
  }

  componentWillUnmount() {
    this.destroy();
  }

  create() {
    const { onChange } = this.props;
    let opts = Object.create(this.props);
    opts['element'] = this.refs.text;
    opts['autoDownloadFontAwesome'] = false;
    opts['spellChecker'] = false;
    let toolbarIcons = [
      'bold',
      'italic',
      'heading',
      '|',
      'code',
      'quote',
      'unordered-list',
      'ordered-list',
      '|',
      'link',
      'image',
      '|',
      'preview',
      'side-by-side',
      'fullscreen',
      '|',
    ];
    opts['toolbar'] = toolbarIcons;
    const editor = new SimpleMDE(opts);
    if (editor.codemirror) {
      editor.codemirror.on('change', () => {
        onChange(editor.value());
      });
    }
    this.editor = editor;
  }

  destroy() {
    for (let i in classNames) {
      let elementToRemove = this.refs.container.querySelector(
        '.' + classNames[i]
      );
      elementToRemove && elementToRemove.remove();
    }
  }

  render() {
    const { name } = this.props;
    return React.createElement(
      'div',
      { ref: 'container' },
      React.createElement('textarea', { ref: 'text', name })
    );
  }
}

MarkdownEditor.propTypes = {
  initialValue: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  onSave: PropTypes.func.isRequired,
};

export default MarkdownEditor;
