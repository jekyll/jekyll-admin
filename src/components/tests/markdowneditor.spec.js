import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import MarkdownEditor from '../MarkdownEditor';

const props = {
  onChange: f => f,
  onSave: f => f,
  placeholder: 'Test',
  initialValue: '',
};

describe('Components::MarkdownEditor', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<MarkdownEditor {...props} />, div);
  });

  it('renders correctly', () => {
    const tree = renderer.create(<MarkdownEditor {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
