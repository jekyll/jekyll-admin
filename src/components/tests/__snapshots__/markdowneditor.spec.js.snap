// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components::MarkdownEditor renders correctly 1`] = `
<div>
  <textarea />
</div>
`;
